ps_emailalerts.php

```diff
@@ -182,7 +185,7 @@ public function uninstallPrestaShop16Module()
                 return parent::uninstall();
             };
             $parentUninstallClosure = $parentUninstallClosure->bindTo($oldModule, get_class($oldModule));
-            $parentUninstallClosure();
+            return $parentUninstallClosure();
```

Comment:
I think this is the only needed change no?

Reply:
No, look at current implementation: 
```
if ($delete_params && $this->uninstallPrestaShop16Module()) {
```

If `$this->uninstallPrestaShop16Module()` is false we never go into the `if` and we never create the SQL table needed for this module.

And in `function uninstallPrestaShop16Module()`
```
        if (!Module::isInstalled(self::PS_16_EQUIVALENT_MODULE)) {
            return false;
        }
```
If the module is installed on PS 17, we just return false. So currently **if you install v2.2.0 on a PS17 store, you dont have the SQL table**. You only have the SQL table if you update the module and there was the PS16 module installed before.

Reply:
> If `$this->uninstallPrestaShop16Module()` is false we never go into the `if` and we never create the SQL table needed for this module.

This is normal, the install should be aborted if the previous module exists AND it cannot be uninstalled. Previously, `$this->uninstallPrestaShop16Module()` would return `true` if it wasn't installed.

Reply:
> This is normal, the install should be aborted if the previous module exists AND it cannot be uninstalled
Yes
Previously, `$this->uninstallPrestaShop16Module()` would return `true` if it wasn't installed.
Not in v2.2.0
https://github.com/PrestaShop/ps_emailalerts/blob/17809a752137a9ec2852c2ee2846f509b81e0d7e/ps_emailalerts.php#L172
```
public function uninstallPrestaShop16Module()
    {
        if (!Module::isInstalled(self::PS_16_EQUIVALENT_MODULE)) {
            return false;
        }
```
This is the key issue that I fix in this PR (or I am missing something big 🤔 )

Reply:
To be honest, the _real_ issue is that we use a boolean to return a 3-states status from `uninstallPrestaShop16Module()`

Status 1: "module 1.6 is not installed"
➡️  then install should continue
Status 2: "module 1.6 installed, but failed to uninstall"
➡️  then install should stop
Status 3: "module 1.6 installed, successfully uninstalled"
➡️  then install should continue

Reply:
> No, look at current implementation:
> 
> ```
> if ($delete_params && $this->uninstallPrestaShop16Module()) {
> ```
> 
> If `$this->uninstallPrestaShop16Module()` is false we never go into the `if` and we never create the SQL table needed for this module.

Yes, this is the desired behavior as @eternoendless said. 


> 
> And in `function uninstallPrestaShop16Module()`
> 
> ```
>         if (!Module::isInstalled(self::PS_16_EQUIVALENT_MODULE)) {
>             return false;
>         }
> ```
> 
> If the module is installed on PS 17, we just return false. So currently **if you install v2.2.0 on a PS17 store, you dont have the SQL table**. You only have the SQL table if you update the module and there was the PS16 module installed before.

You are wrong here, the original statement is 
```
        if (!Module::isInstalled(self::PS_16_EQUIVALENT_MODULE)) {
            return true;
        }
```

which is returning true :) 

We can interpret the `uninstallPrestaShop16Module` returns as `@return true if everything is fine`

Reply:
You can always rename it to `uninstallPrestaShop16ModuleIfNecessary()`